text/0000-variadic-generics.md

```diff
@@ -0,0 +1,207 @@
+    fn map(self, _: F) {}
+}
+
+impl<Head, Tail, F, R> TupleMap<F> for (Head, ...Tail)
+    where
+    F: Fn(Head) -> R,
+    Tail: TupleMap<F>,
+{
+    type Out = (R, ...<Tail as TupleMap<F>>::Out);
+    
+    fn map(self, f: F) -> Self::Out {
+        let (head, ...tail) = self;
+        (f(head), ...tail.map(f))
+    }
+}
+```
+
+This example is derived from
+[a playground example by @eddyb](https://play.rust-lang.org/?gist=8fd29c83271f3e8744a3f618786ca1de&version=nightly&backtrace=0)
+that provided inspiration for this RFC.
+
+The example demonstrates the concise, expressive code enabled
+by this RFC. In order to implement a trait for tuples of any length, all
+that was necessary was to implement the trait for `()` and `(Head, ...Tail)`.
+
+# How We Teach This
+[teach]: #teach
+
+The `(head, ...tail)` and `(Head, ...Tail)` syntax closely mirror established
+patterns for working with `Cons`-cell based lists. Rustaceans coming from
+other functional programming languages will likely be familiar with the concept
+of recursively-defined lists. For those unfamiliar with `Cons`-based
+lists, the concept should be introduced using "structural recursion": there's
+a base case, `()`, and a recursive/inductive case: `(Head, ...Tail)`. Any tuple
+can be thought of in this way
+(for example, `(A, B, C)` is equivalent to `(A, ...(B, ...(C, ...())))`).
+
+The exact mechanisms used to teach this should be determined after getting more
+experience with how Rustaceans learn. After all, Rust users are a diverse crowd,
+so the "best" way to teach one person might not work as well for another. There
+will need to be some investigation into which explanations are more
+suitable to a general audience.
+
+As for the `(head, ...tail)` joining syntax, this should be explained as
+taking each part of the tail (e.g. `(2, 3, 4)`) and inlining or un-"tupling"
+them (e.g. `2, 3, 4`). This is nicely symmetrical with the `(head, ...tail)`
+pattern-matching syntax.
+
+The `Tuple` trait is a bit of an oddity. It is probably best not to go too
+far into the weeds when explaining it to new users. The extra coherence
+benefits will likely go unnoticed by new users, as they allow for more
+advanced features and wouldn't result in an error where one didn't exist
+before. The obvious exception is when trying to implement the `Tuple` trait.
+Attempts to implement `Tuple` should resort in a relevant error message,
+such as "The `Tuple` trait cannot be implemented for custom types."
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+As with any additions to the language, this RFC would increase the number
+of features present in Rust, potentially resulting increased complexity
+of the language.
+
+There is also some unfortunate overlap between the proposed `(head, ...tail)`
+syntax and the current inclusive range syntax. However, the similarity
+between `start...end` and `...tail` can be disambiguiated by whether or not
+there is an expression immediately before the ellipsis.
```

Comment:
Yes, you're correct.  @eddyb [made the same observation above](https://github.com/rust-lang/rfcs/pull/1935#discussion_r103383894).